// electron-builder/out/targets/ArchiveTarget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ArchiveTarget = undefined;

var _bluebirdLst;

function _load_bluebirdLst() {
    return _bluebirdLst = require("bluebird-lst");
}

var _builderUtil;

function _load_builderUtil() {
    return _builderUtil = require("builder-util");
}

var _path = _interopRequireWildcard(require("path"));

var _core;

function _load_core() {
    return _core = require("../core");
}

var _archive;

function _load_archive() {
    return _archive = require("./archive");
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

class ArchiveTarget extends (_core || _load_core()).Target {
    constructor(name, outDir, packager, isWriteUpdateInfo = false) {
        super(name);
        this.outDir = outDir;
        this.packager = packager;
        this.isWriteUpdateInfo = isWriteUpdateInfo;
        this.options = this.packager.config[this.name];
    }
    build(appOutDir, arch) {
        var _this = this;

        return (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* () {
            const packager = _this.packager;
            const isMac = packager.platform === (_core || _load_core()).Platform.MAC;
            const format = _this.name;
            (0, (_builderUtil || _load_builderUtil()).log)(`Building ${isMac ? "macOS " : ""}${format}`);
            let defaultPattern;
            if (packager.platform === (_core || _load_core()).Platform.LINUX) {
                // tslint:disable-next-line:no-invalid-template-strings
                defaultPattern = "${name}-${version}" + (arch === (_builderUtil || _load_builderUtil()).Arch.x64 ? "" : "-${arch}") + ".${ext}";
            } else {
                // tslint:disable-next-line:no-invalid-template-strings
                defaultPattern = "${productName}-${version}" + (arch === (_builderUtil || _load_builderUtil()).Arch.x64 ? "" : "-${arch}") + "-${os}.${ext}";
            }
            const artifactPath = _path.join(_this.outDir, packager.expandArtifactNamePattern(_this.options, format, arch, defaultPattern, false));
            if (format.startsWith("tar.")) {
                yield (0, (_archive || _load_archive()).tar)(packager.compression, format, artifactPath, appOutDir, isMac, packager.info.tempDirManager);
            } else {
                yield (0, (_archive || _load_archive()).archive)(packager.compression, format, artifactPath, appOutDir, { withoutDir: !isMac });
            }
            packager.info.dispatchArtifactCreated({
                file: artifactPath,
                safeArtifactName: isMac ? packager.generateName2(format, "mac", true) : packager.generateName(format, arch, true, packager.platform === (_core || _load_core()).Platform.WINDOWS ? "win" : null),
                target: _this,
                arch,
                packager,
                isWriteUpdateInfo: _this.isWriteUpdateInfo
            });
        })();
    }
}
exports.ArchiveTarget = ArchiveTarget; //# sourceMappingURL=ArchiveTarget.js.map
