// @babel/core/lib/config/build-config-chain.js
"use strict";

exports.__esModule = true;
exports.default = buildConfigChain;

var _environment = require("./helpers/environment");

var _path = _interopRequireDefault(require("path"));

var _micromatch = _interopRequireDefault(require("micromatch"));

var _debug = _interopRequireDefault(require("debug"));

var _options = require("./options");

var _files = require("./loading/files");

var _caching = require("./caching");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

var debug = (0, _debug.default)("babel:config:config-chain");

function buildConfigChain(opts) {
  var filename = opts.filename ? _path.default.resolve(opts.filename) : null;
  var builder = new ConfigChainBuilder(filename ? new LoadedFile(filename) : null);
  var envKey = (0, _environment.getEnv)();

  try {
    builder.mergeConfigArguments(opts, process.cwd(), envKey);

    if (opts.babelrc !== false && filename) {
      (0, _files.findConfigs)(_path.default.dirname(filename)).forEach(function (configFile) {
        return builder.mergeConfigFile(configFile, envKey);
      });
    }
  } catch (e) {
    if (e.code !== "BABEL_IGNORED_FILE") throw e;
    return null;
  }

  return builder.configs.reverse();
}

var ConfigChainBuilder = function () {
  function ConfigChainBuilder(file) {
    this.file = void 0;
    this.configs = [];
    this.seenFiles = new Set();
    this.file = file;
  }

  var _proto = ConfigChainBuilder.prototype;

  _proto.mergeConfigArguments = function mergeConfigArguments(opts, dirname, envKey) {
    var _this = this;

    flattenArgumentsOptionsParts(opts, dirname, envKey).forEach(function (part) {
      return _this._processConfigPart(part, envKey);
    });
  };

  _proto.mergeConfigFile = function mergeConfigFile(file, envKey) {
    var _this2 = this;

    if (this.seenFiles.has(file)) {
      throw new Error("Cycle detected in Babel configuration file through \"" + file.filepath + "\".");
    }

    var parts = flattenFileOptionsParts(file)(envKey);
    this.seenFiles.add(file);
    parts.forEach(function (part) {
      return _this2._processConfigPart(part, envKey);
    });
    this.seenFiles.delete(file);
  };

  _proto._processConfigPart = function _processConfigPart(part, envKey) {
    if (part.part === "config") {
      var _ignore = part.ignore,
          _only = part.only;

      if (this.file && this.file.shouldIgnore(_ignore, _only, part.config.dirname)) {
        throw Object.assign(new Error("This file has been ignored."), {
          code: "BABEL_IGNORED_FILE"
        });
      }

      this.configs.push(part.config);
    } else {
      this.mergeConfigFile((0, _files.loadConfig)(part.path, part.dirname), envKey);
    }
  };

  return ConfigChainBuilder;
}();

function flattenArgumentsOptionsParts(opts, dirname, envKey) {
  var env = opts.env,
      plugins = opts.plugins,
      presets = opts.presets,
      passPerPreset = opts.passPerPreset,
      extendsPath = opts.extends,
      options = _objectWithoutProperties(opts, ["env", "plugins", "presets", "passPerPreset", "extends"]);

  var raw = [];

  if (env) {
    raw.push.apply(raw, flattenArgumentsEnvOptionsParts(env)(dirname)(envKey));
  }

  if (Object.keys(options).length > 0) {
    raw.push.apply(raw, flattenOptionsParts(buildArgumentsItem(options, dirname)));
  }

  if (plugins) {
    raw.push.apply(raw, flattenArgumentsPluginsOptionsParts(plugins)(dirname));
  }

  if (presets) {
    raw.push.apply(raw, flattenArgumentsPresetsOptionsParts(presets)(!!passPerPreset)(dirname));
  }

  if (extendsPath != null) {
    raw.push.apply(raw, flattenOptionsParts(buildArgumentsItem({
      extends: extendsPath
    }, dirname)));
  }

  return raw;
}

var flattenArgumentsEnvOptionsParts = (0, _caching.makeWeakCache)(function (env) {
  var options = {
    env: env
  };
  return (0, _caching.makeStrongCache)(function (dirname) {
    return flattenOptionsPartsLookup(buildArgumentsItem(options, dirname));
  });
});
var flattenArgumentsPluginsOptionsParts = (0, _caching.makeWeakCache)(function (plugins) {
  var options = {
    plugins: plugins
  };
  return (0, _caching.makeStrongCache)(function (dirname) {
    return flattenOptionsParts(buildArgumentsItem(options, dirname));
  });
});
var flattenArgumentsPresetsOptionsParts = (0, _caching.makeWeakCache)(function (presets) {
  return (0, _caching.makeStrongCache)(function (passPerPreset) {
    var options = {
      presets: presets,
      passPerPreset: passPerPreset
    };
    return (0, _caching.makeStrongCache)(function (dirname) {
      return flattenOptionsParts(buildArgumentsItem(options, dirname));
    });
  });
});

function buildArgumentsItem(options, dirname) {
  return {
    type: "arguments",
    options: options,
    alias: "base",
    dirname: dirname
  };
}

var flattenFileOptionsParts = (0, _caching.makeWeakCache)(function (file) {
  return flattenOptionsPartsLookup({
    type: "file",
    options: (0, _options.validate)("file", file.options),
    alias: file.filepath,
    dirname: file.dirname
  });
});

function flattenOptionsPartsLookup(config) {
  var parts = flattenOptionsParts(config);
  var def = parts.filter(function (part) {
    return part.activeEnv === null;
  });
  var lookup = new Map();
  parts.forEach(function (part) {
    if (part.activeEnv !== null) lookup.set(part.activeEnv, []);
  });

  var _loop = function _loop(values, _activeEnv) {
    parts.forEach(function (part) {
      if (part.activeEnv === null || part.activeEnv === _activeEnv) {
        values.push(part);
      }
    });
  };

  for (var _iterator = lookup, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
    var _ref2;

    if (_isArray) {
      if (_i >= _iterator.length) break;
      _ref2 = _iterator[_i++];
    } else {
      _i = _iterator.next();
      if (_i.done) break;
      _ref2 = _i.value;
    }

    var _ref3 = _ref2;
    var _activeEnv = _ref3[0];
    var values = _ref3[1];

    _loop(values, _activeEnv);
  }

  return function (envKey) {
    return lookup.get(envKey) || def;
  };
}

function flattenOptionsParts(config, activeEnv) {
  if (activeEnv === void 0) {
    activeEnv = null;
  }

  var rawOpts = config.options,
      alias = config.alias,
      dirname = config.dirname;
  var parts = [];

  if (rawOpts.env) {
    var _arr = Object.keys(rawOpts.env);

    for (var _i2 = 0; _i2 < _arr.length; _i2++) {
      var envKey = _arr[_i2];

      if (rawOpts.env[envKey]) {
        parts.push.apply(parts, flattenOptionsParts({
          type: "env",
          options: rawOpts.env[envKey],
          alias: alias + (".env." + envKey),
          dirname: dirname
        }, envKey));
      }
    }
  }

  parts.push({
    part: "config",
    config: config,
    ignore: rawOpts.ignore,
    only: rawOpts.only,
    activeEnv: activeEnv
  });

  if (rawOpts.extends != null) {
    parts.push({
      part: "extends",
      path: rawOpts.extends,
      dirname: dirname,
      activeEnv: activeEnv
    });
  }

  return parts;
}

var LoadedFile = function () {
  function LoadedFile(filename) {
    this.filename = void 0;
    this.possibleDirs = null;
    this.filename = filename;
  }

  var _proto2 = LoadedFile.prototype;

  _proto2.shouldIgnore = function shouldIgnore(ignore, only, dirname) {
    if (ignore) {
      if (this._matchesPatterns(ignore, dirname)) {
        debug("Ignored %o because it matched one of %O from %o", this.filename, ignore, dirname);
        return true;
      }
    }

    if (only) {
      if (!this._matchesPatterns(only, dirname)) {
        debug("Ignored %o because it failed to match one of %O from %o", this.filename, only, dirname);
        return true;
      }
    }

    return false;
  };

  _proto2._matchesPatterns = function _matchesPatterns(patterns, dirname) {
    var res = [];
    var strings = [];
    var fns = [];
    patterns.forEach(function (pattern) {
      if (typeof pattern === "string") strings.push(pattern);else if (typeof pattern === "function") fns.push(pattern);else res.push(pattern);
    });
    var filename = this.filename;
    if (res.some(function (re) {
      return re.test(filename);
    })) return true;
    if (fns.some(function (fn) {
      return fn(filename);
    })) return true;

    if (strings.length > 0) {
      var possibleDirs = this.possibleDirs;

      if (!possibleDirs) {
        possibleDirs = this.possibleDirs = [];
        possibleDirs.push(filename);
        var current = filename;

        while (true) {
          var previous = current;
          current = _path.default.dirname(current);
          if (previous === current) break;
          possibleDirs.push(current);
        }
      }

      var absolutePatterns = strings.map(function (pattern) {
        var negate = pattern[0] === "!";
        if (negate) pattern = pattern.slice(1);
        return (negate ? "!" : "") + _path.default.resolve(dirname, pattern);
      });

      if ((0, _micromatch.default)(possibleDirs, absolutePatterns, {
        nocase: true
      }).length > 0) {
        return true;
      }
    }

    return false;
  };

  return LoadedFile;
}();
